Mage.Sets/src/mage/cards/k/KayaSpiritsJustice.java

```diff
@@ -0,0 +1,266 @@
+                .filter(e -> e.getToZone() == Zone.EXILED)
+                .map(ZoneChangeEvent::getTargetId)
+                .filter(Objects::nonNull)
+                .map(game::getCard)
+                .filter(Objects::nonNull)
+                .filter(card -> {
+                    Permanent permanent = game.getPermanentOrLKIBattlefield(card.getId());
+                    return StaticFilters.FILTER_PERMANENT_CREATURE
+                            .match(permanent, getControllerId(), this, game);
+                })
+                .collect(Collectors.toSet());
+
+        Set<Card> graveyardCards = zEvent.getEvents()
+                .stream()
+                .filter(e -> e.getFromZone() == Zone.GRAVEYARD)
+                .filter(e -> e.getToZone() == Zone.EXILED)
+                .map(ZoneChangeEvent::getTargetId)
+                .filter(Objects::nonNull)
+                .map(game::getCard)
+                .filter(Objects::nonNull)
+                .filter(card -> StaticFilters.FILTER_CARD_CREATURE
+                        .match(card, getControllerId(), this, game))
+                .collect(Collectors.toSet());
+
+        Set<Card> cards = new HashSet<>(battlefieldCards);
+        cards.addAll(graveyardCards);
+        if (cards.isEmpty()) {
+            return false;
+        }
+
+        getEffects().setTargetPointer(new FixedTargets(new CardsImpl(cards), game));
+        return true;
+    }
+}
+
+class KayaSpiritsJusticeCopyEffect extends OneShotEffect {
+
+    KayaSpiritsJusticeCopyEffect() {
+        super(Outcome.Copy);
+    }
+
+    private KayaSpiritsJusticeCopyEffect(final KayaSpiritsJusticeCopyEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public KayaSpiritsJusticeCopyEffect copy() {
+        return new KayaSpiritsJusticeCopyEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        Permanent copyToPermanent = game.getPermanent(source.getFirstTarget());
+        Cards exiledCards = new CardsImpl(getTargetPointer().getTargets(game, source));
+        if (controller == null || copyToPermanent == null || exiledCards.isEmpty()) {
+            return false;
+        }
+
+        TargetCard target = new TargetCardInExile(0, 1, StaticFilters.FILTER_CARD_CREATURE, null);
+        if (!target.canChoose(source.getControllerId(), source, game)) {
+            return false;
+        }
+        if (!controller.chooseTarget(outcome, exiledCards, target, source, game)) {
+            return false;
+        }
+
+        Card copyFromCard = game.getCard(target.getFirstTarget());
+        if (copyFromCard == null) {
+            return false;
+        }
+
+        game.addEffect(new CopyEffect(Duration.EndOfTurn, copyFromCard, copyToPermanent.getId()), source);
+        game.addEffect(new GainAbilityTargetEffect(FlyingAbility.getInstance(), Duration.EndOfTurn), source);
```

Comment:
It’s part of copy, must use CopyApplier instead gain.

Example:
![IMG_9622](https://github.com/magefree/mage/assets/8344157/01258b8e-4c4d-44c5-9505-a49862d6ca64)